### PR TITLE
re-implement updation of CoordinatorDynamicConfig to not call manager.watch(..) in its constructor

### LINKED ISSUE #4555: Bug in CoordinatorDynamicConfig
https://github.com/druid-io/druid/pull/4141 added watcher to be added as part of CoordinatorDynamicConfig construction which can lead to a deadlock in ConfigManager due to single threaded "exec" pool in ConfigManager.

### PR DESCRIPTION
**Pull request by himanshug**
effectively reverts #4141 and leaves the responsibility of providing correctly updated CoordinatorDynamicConfig to CoordinatorDynamicConfigResource

Fixes #4555

### CODE REVIEW

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 1]
```diff
@@ -428,17 +373,34 @@ public Builder withMaxSegmentsInNodeLoadingQueue(int maxSegmentsInNodeLoadingQue
...
       return new CoordinatorDynamicConfig(
-          millisToWaitBeforeDeleting,
-          mergeBytesLimit,
-          mergeSegmentsLimit,
-          maxSegmentsToMove,
-          replicantLifetime,
-          replicationThrottleLimit,
-          balancerComputeThreads,
-          emitBalancingStats,
+          millisToWaitBeforeDeleting == null ? 15 * 60 * 1000L : millisToWaitBeforeDeleting,
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Please use `TimeUnit.MINUTES.toMillis(15)`

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java [hunk 2]
```diff
@@ -428,17 +373,34 @@ public Builder withMaxSegmentsInNodeLoadingQueue(int maxSegmentsInNodeLoadingQue
...
+          mergeBytesLimit == null ? 524288000L : mergeBytesLimit,
```
Comment (leventov) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: Please extract all those magic numbers as constants.

File: server/src/main/java/io/druid/server/http/CoordinatorDynamicConfigsResource.java [hunk 3]
```diff
@@ -76,15 +76,20 @@ public Response getDynamicConfigs()
...
   {
+    CoordinatorDynamicConfig current = manager.watch(
+        CoordinatorDynamicConfig.CONFIG_KEY,
+        CoordinatorDynamicConfig.class
+    ).get();
+
     if (!manager.set(
         CoordinatorDynamicConfig.CONFIG_KEY,
-        dynamicConfig,
+        current == null ? dynamicConfigBuilder.build() : dynamicConfigBuilder.build(current),
```
Comment (leventov) [inline server/src/main/java/io/druid/server/http/CoordinatorDynamicConfigsResource.java]: Probably vise versa

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 1 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: changed

File: server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java (hunk 2 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java]: done

File: server/src/main/java/io/druid/server/http/CoordinatorDynamicConfigsResource.java (hunk 3 above)
Comment (himanshug) [inline server/src/main/java/io/druid/server/http/CoordinatorDynamicConfigsResource.java]: ordering is correct. "current" is what is stored in DB now ... so if there is nothing in db (current == null case) then use generic defaults or use values in "current" as defaults.

**Comment by pjain1** [comment]
👍

**Comment by himanshug** [comment]
@leventov @jon-wei  TeamCity build failed in this PR , I looked at the logs and errors did not look related to updates made in this PR. Is that expected to fail for now ?

**Comment by gianm** [comment]
I've noticed similar issues on other patches (teamcity failing on files unrelated to the patch). It seems flaky for some reason.

**Comment by himanshug** [comment]
ok, backporting it to 0.10.1 now

**Pull request merged**